### PR TITLE
Feature: support discriminators

### LINKED ISSUE #36: getDiscriminatorModelForClass() is not being used in case where models are sub-typed
@kpfromer Refer details here -- https://typegoose.github.io/typegoose/docs/functions/getDiscriminatorModelForClass/.

With all models created using `getModelForClass` method, `getDiscriminatorModelForClass` is not being used and the discriminator key in schema-options is not being set as a result.  I think the initialization of the model should consider a discriminator when a model is being instantiated.

### PR DESCRIPTION
**Pull request by beaulac**
Closes #36 

Adds support for discriminators by allowing users to specify them in model options. This adds a `discriminators` array property to the long-form model options. 

The `discriminators` array can consist of either bare classes, or an object similar to model options that allows overriding the discriminator key value.

Here is the example from the updated README:
```typescript
class Tabby extends Cat {
  @prop()
  spotted: boolean 
}

class BlackCat extends Cat {
  @prop()
  unlucky: boolean
}

@Module({
  imports: [
    TypegooseModule.forFeature([
      {
        typegooseClass: Cat,
        discriminators: [
          Tabby,
          {
            typegooseClass: BlackCat,
            discriminatorId: 'Black'
          }
        ]
      }
    ])
  ]
})
export class CatsModule {}
```

Discriminator models are provided & injected like any other model.

Using an internal map was suggested in #36, but Typegoose uses an internal map for models, so `getModelForClass` and `getDiscriminatorModelForClass` are idempotent. The approach used to create these models makes the discriminator model providers keep a reference to their parent's factory.

As far as the tests show, this change maintains backwards compatibility.

Let me know if anything needs to be changed, and thanks for this great library.

### CODE REVIEW

**Comment by hasezoey** [comment]
Note: current travis fail is because this project still uses typescript < 3.5, thanks to `Omit`

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "nestjs-typegoose",
-  "version": "7.0.0",
+  "version": "7.1.0",
```
Comment (hasezoey) [inline package.json]: i would recommend to revert this, the maintainer @kpfromer should only do this

Comment (beaulac) [reply package.json]: OK!

I was following the procedure outlined here: https://github.com/kpfromer/nestjs-typegoose/blob/master/CONTRIBUTING.md#pull-request-process

If PRs should not update version, perhaps that instruction should be removed from the contribution guidelines.

Comment (hasezoey) [reply package.json]: OK nevermind, didnt read that, sorry - i thought it was "common knowledge" to not increment it, when not being able to release it (thanks to something like `npm version *` scripts)

Comment (hasezoey) [reply package.json]: otherwise, to comply with semver:
because this is not a breaking feature, i would recommend only increment the patch number

Comment (beaulac) [reply package.json]: Hm, maybe that is from a different semver spec?

From https://semver.org/ , both v1.0.0 and v2.0.0 specs have a line starting with
> Minor version Y (x.Y.z | x > 0) MUST be incremented if new, backwards compatible functionality is introduced to the public API.

Comment (hasezoey) [reply package.json]: ok, i thought about npm-semver https://docs.npmjs.com/about-semantic-versioning#incrementing-semantic-versions-in-published-packages

Comment (hasezoey) [reply package.json]: nvm confused "Backward compatible ***bug fixes***" with "Backward compatible ***new features***"

Comment (hasezoey) [reply package.json]: consider this resolved (cant resolve this)

File: src/typegoose.providers.ts [hunk 2]
```diff
@@ -1,36 +1,79 @@
...
+    (providers, { typegooseClass, schemaOptions = {}, discriminators = [] }) => {
+
+      const modelFactory = (connection: Connection) => getModelForClass(
+        typegooseClass,
+        { existingConnection: connection, schemaOptions },
+      );
+
+      const modelProvider = buildProvider(typegooseClass, modelFactory);
+
+      const discriminatorProviders = discriminators.map(createDiscriminatorFactoryFrom(modelFactory));
```
Comment (kpfromer) [inline src/typegoose.providers.ts]: `createDiscriminatorFactoryFrom` can only handle a `TypegooseDiscriminator` but here it is provided a union of `TypegooseDiscriminator | TypegooseClass`.

Comment (beaulac) [reply src/typegoose.providers.ts]: Nice catch! Not sure how it was able to compile / pass tests before, though 🤔

Comment (kpfromer) [reply src/typegoose.providers.ts]: Yeah, it was really weird when I was updating to master's changes and typescript then decided to error. I did push a commit to fix it though.

**Pull request merged**